cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+Existing mutual authentication implementations put several restrictions upon users in order to attain additional security guarantees. These can range from requiring apps to use TCP+TLS for authentication/encryption to requiring (sidecar) proxies in the dataplane. These restrictions increase the baseline complexity and processing cost of implementing mutual authentication. We believe that by combining the best of session-based authentication and network-based authentication, we can provide a secure implementation that is faster and more flexible than prior implementations.
+
+## Goals
+
+* Provide a mechanism for pods to opt into peer authentication
+* Pluggable for existing certificate management systems (SPIFFE, Vault, SMI, Istio, cert-manager, …)
+* Configurable certificate granularity
+* Leverage existing encryption protocol support for the dataplane
+
+
+## Non-Goals
+
+* Inline authentication via a proxy
+* Drop less connection establishment
+
+
+## Proposal
+
+### Overview
+
+The Isovalent blog Next-Generation Mutual Authentication with Cilium Service Mesh described the high level design for mutual authentication where there are two channels for connectivity: Firstly, a control plane connection between cilium-agent instances on each node provides authentication of connections between pods on the nodes. Secondly, the existing Cilium encryption support using WireGuard or IPsec provides an encrypted dataplane for the connections. This document seeks to describe the per-node implementation that coordinates these out-of-band authentication channels with the in-band network encryption channels.
+
+Here is a diagram showing the key parts of the design
+
+![mTLS General Architecture](./images/mTLS-general-architecture.png)
+
+The following sub-sections are broken up by the aspects of the design where specific changes need to be made, starting from the k8s resource types that need to be modified, and followed by the considerations for the cert management control plane, the (local) control plane, and the data plane. While the following sections may focus more on SPIFFE as a default auth implementation for the pluggable interface, the same principles should apply for other auth mechanisms.
+
+### Resources
+
+TODO: This section needs an update.
+
+#### ConfigMap - cilium-config
+
+* option-name: `default`
+  * description
+* mesh-auth-mode: `disabled`
+  * enabled / disabled
+* mesh-auth-cert-source: `spiffe`
+  * Configure certificate mechanism, default to directory path
+* mesh-spire-server: `____`
+  * Configure the spire server
+* mesh-spire-agent: `unix://var/run/cilium/spiffe/agent.sock`
+  * Configure the SPIRE agent
+* mesh-auth-path: `/var/run/cilium/auth/`
+  * Configure the directory to pull certificates from, if mesh-auth-cert is directory
```

Comment:
Is directory going to be one of the options? I seem to have missed a mention above?

Reply:
No, this should have been removed, thanks!